### PR TITLE
Add dependency to the running_deps

### LINKED ISSUE #20461: Airflow is trying to schedule tasks prior to DAG's start_date
### Apache Airflow version

2.2.2

### What happened

I have a DAG which start_date was 09/01/2021 initially but then it was changed to 11/01/2021. This DAG has some runs prior to 11/01/2021 that did not get a chance to finish. I can now see that scheduler is still trying to schedule the runs prior to 11/01/2021. But none of the tasks in these runs are starting because of the start_date check I presume. This is maxing out the active runs thus blocking any other days within the DAG range to be scheduled.

Screen Shot 2021-12-22 at 7 56 22 AM
Screen Shot 2021-12-22 at 7 56 29 AM

### What you expected to happen

Scheduler should not be trying to schedule runs that are prior to DAG's start_date.

### How to reproduce

_No response_

### Operating System

Airflow Docker

### Versions of Apache Airflow Providers

_No response_

### Deployment

Other Docker-based deployment

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by avkirilishin**
closes: #20461

Problem: Airflow is trying to schedule tasks prior to DAG's start_date.
Solution: Added ExecDateAfterStartDateDep to the dependencies that need to be met for a given task instance to be set to 'RUNNING' state.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/ti_deps/deps/exec_date_after_start_date_dep.py [hunk 1]
```diff
@@ -28,15 +28,19 @@ class ExecDateAfterStartDateDep(BaseTIDep):
 
     @provide_session
     def _get_dep_statuses(self, ti, session, dep_context):
-        if ti.task.start_date and ti.execution_date < ti.task.start_date:
+        if ti.task.start_date and ti.get_dagrun(session=session).execution_date < ti.task.start_date:
```
Comment (ephraimbuddy) [inline airflow/ti_deps/deps/exec_date_after_start_date_dep.py:31]: Any reason why we are changing this? I think the ti.execution_date will be available at this point.

File: tests/models/test_dag.py [hunk 2]
```diff
@@ -2029,14 +2029,14 @@ def test_count_number_queries(self, tasks_count):
 
 
 class TestDagDecorator(unittest.TestCase):
+    DEFAULT_DATE = timezone.datetime(2016, 1, 1)
     DEFAULT_ARGS = {
         "owner": "test",
         "depends_on_past": True,
-        "start_date": timezone.utcnow(),
+        "start_date": DEFAULT_DATE,
```
Comment (ephraimbuddy) [inline tests/models/test_dag.py:2036]: Unrelated?

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
I don't think this is the right fix. If I understand correctly, the problem is that the scheduler is creating dagruns for those past dates and not that the task instances are being scheduled. The ExecDateAfterStartDateDep currently in place will prevent tasks from getting into queued state if the start date is before the execution date

**Comment by avkirilishin** [comment]
> I don't think this is the right fix. If I understand correctly, the problem is that the scheduler is creating dagruns for those past dates and not that the task instances are being scheduled. The ExecDateAfterStartDateDep currently in place will prevent tasks from getting into queued state if the start date is before the execution date

I explained it here: https://github.com/apache/airflow/issues/20461#issuecomment-1046051990

> I think there are two different problems:
>
> The problem is related to the different logic of the scheduler before and after the update. Maybe there are no tasks in the running dags or something else. Can you show the rows for this dag run in dag, dag_run and task_instance?
> 
> _**I agree with you that it is not the right behavior for the scheduler to continue scheduling runs that are earlier than the actual start date of the DAG or Task. It can happen, for example, after turning the dag off and back on. So I made a PR to fix it: https://github.com/apache/airflow/pull/21684**_

File: airflow/ti_deps/deps/exec_date_after_start_date_dep.py (hunk 1 above)
Comment (avkirilishin) [reply airflow/ti_deps/deps/exec_date_after_start_date_dep.py:31]: After your commit https://github.com/apache/airflow/commit/80ae70c3d0f6109e974d50726da5a44e58d1b674 DagRun must be loaded first (via get_dagrun) or TaskInstance's execution_date would be None

File: tests/models/test_dag.py (hunk 2 above)
Comment (avkirilishin) [reply tests/models/test_dag.py:2036]: It affects `test_dag_param_dagrun_parameterized` and `test_dag_param_resolves` where `execution_date` is `self.DEFAULT_DATE` (2016-01-01), but `execution_date` can not be less than DAG's start_date.
